command/sub8_missions/missions/start_gate.py

```diff
@@ -1,118 +1,145 @@
 #!/usr/bin/env python
 from __future__ import division
 
-# from txros import action, util, tf, serviceclient
 import txros
+import txros.tf as tf
 
 from mil_misc_tools import text_effects
-from sub8_msgs.srv import VisionRequest2DRequest, VisionRequest2D, BMatrix, BMatrixRequest
-from std_srvs.srv import SetBool, SetBoolRequest
+from geometry_msgs.msg import Point, Vector3, Pose
 
-import numpy as np
 
 fprint = text_effects.FprintFactory(
     title="START_GATE", msg_color="white").fprint
 
 
 class StartGateMission(object):
-
-    FOUND_START_GATE = False
     SPEED = 0.3
+    HOW_MANY = 20
 
     def __init__(self, sub):
-        self.sub_singleton = sub
+        # TODO: vision proxy
+        self._sub = sub
+        self._how_many_points = 0
+        self._how_many_normals = 0
+        self._center_pt_avg = Point()
+        self._normal_avg = Vector3()
+        self._found_center_pt = False
+        self._found_normal = False
+        self._found = False
+
+        self._tf_listener = self._sub._tf_listener
+
+        self._start_away_from_gate = 1
+        self._end_away_from_gate = 0.5
+
+        self._center_of_gate_sub = self._sub.nh.subscribe(
+            '/sub8_start_gate/center', Point, callback=self.process_centers)
+        self._normal_of_gate_sub = self._sub.nh.subscribe(
+            '/sub8_start_gate/normal', Vector3, callback=self.process_normals)
+
+    def process_centers(self, point):
+        if self._found:
+            return
+        self._how_many_points = self._how_many_points + 1
+        self._center_pt_avg.x += point.x
+        self._center_pt_avg.y += point.y
+        self._center_pt_avg.z += point.z
+
+        if(self._how_many_points > self.HOW_MANY):
+            self._found_center_pt = True
+
+    def process_normals(self, normal):
+        if self._found:
+            return
+        self._how_many_normals = self._how_many_normals + 1
+        if(normal.z > 0):
+            normal.x = -normal.x
+            normal.y = -normal.y
+            normal.z = -normal.z
+        self._normal_avg.x += normal.x
+        self._normal_avg.y += normal.y
+        self._normal_avg.z += normal.z
+
+        if(self._how_many_normals > self.HOW_MANY):
+            self._found_normal = True
 
     @txros.util.cancellableInlineCallbacks
     def run_mission(self):
```

Comment:
This `run_mission` function is pretty long, you should refactor into smaller chunks.